File: android/src/main/java/com/getcapacitor/community/Bridgefy.java
package com.getcapacitor.community;

import android.content.Context;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import com.getcapacitor.community.classes.options.EstablishSecureConnectionOptions;
import com.getcapacitor.community.classes.options.FingerprintOptions;
import com.getcapacitor.community.classes.options.InitializeOptions;
import com.getcapacitor.community.classes.options.IsFingerprintValidOptions;
import com.getcapacitor.community.classes.options.SendOptions;
import com.getcapacitor.community.classes.options.StartOptions;
import com.getcapacitor.community.classes.results.ConnectedPeersResult;
import com.getcapacitor.community.classes.results.FingerprintResult;
import com.getcapacitor.community.classes.results.IsFingerprintValidResult;
import com.getcapacitor.community.classes.results.IsInitializedResult;
import com.getcapacitor.community.classes.results.IsStartedResult;
import com.getcapacitor.community.classes.results.LicenseExpirationDateResult;
import com.getcapacitor.community.classes.results.SendResult;
import com.getcapacitor.community.classes.results.UserIDResult;
import com.getcapacitor.community.interfaces.Callback;
import java.util.Date;
import java.util.List;
import java.util.UUID;

public class Bridgefy {

    private final String MISSING_API_KEY = "missing API key";
    private final String MISSING_USER_IDENTIFIER = "missing user identifier";
    private final String MISSING_FINGERPRINT = "missing fingerprint";
    private final String MISSING_PAYLOAD = "missing payload";
    private final String MISSING_TRANSMISSION_MODE = "missing transmission mode";

    @NonNull
    private final BridgefyConfig config;

    private final BridgefyController bridgefy;

    public Bridgefy(@NonNull BridgefyConfig config, @NonNull BridgefyPlugin plugin) {
        this.config = config;

        bridgefy = new BridgefyController(plugin);
    }

    /**
     * Initialize
     */

    public void initialize(@NonNull InitializeOptions options, @NonNull Callback callback) {
        @Nullable
        UUID apiKey = options.getAPIKey() != null ? options.getAPIKey() : config.getAPIKey();

        if (apiKey == null) {
            callback.error(new Exception(MISSING_API_KEY));
            return;
        }

        @Nullable
        Boolean verboseLogging = options.getVerboseLogging() != null ? options.getVerboseLogging() : config.getVerboseLogging();

        bridgefy.initialize(apiKey, verboseLogging);

        callback.success();
    }

    public void isInitialized(@NonNull Callback callback) {
        Boolean isInitialized = bridgefy.isInitialized();

        IsInitializedResult result = new IsInitializedResult(isInitialized);
        callback.success(result);
    }

    public void start(@NonNull StartOptions options, @NonNull Callback callback) {
        @Nullable
        UUID userID = options.getUserID();

        @Nullable
        String propagationProfile = options.getPropagationProfile();

        bridgefy.start(userID, propagationProfile);

        callback.success();
    }

    public void isStarted(@NonNull Callback callback) {
        Boolean isStarted = bridgefy.isStarted();

        IsStartedResult result = new IsStartedResult(isStarted);
        callback.success(result);
    }

    public void stop(@NonNull Callback callback) {
        bridgefy.stop();

        callback.success();
    }

    /**
     * License
     */

    public void licenseExpirationDate(@NonNull Callback callback) {
        Date licenseExpirationDate = bridgefy.licenseExpirationDate();

        LicenseExpirationDateResult result = new LicenseExpirationDateResult(licenseExpirationDate);
        callback.success(result);
    }

    public void updateLicense(@NonNull Callback callback) {
        bridgefy.updateLicense();

        callback.success();
    }

    /**
     * Session
     */

    public void destroySession(@NonNull Callback callback) {
        bridgefy.destroySession();

        callback.success();
    }

    public void currentUserID(@NonNull Callback callback) {
        UUID userID = bridgefy.currentUserId();

        UserIDResult result = new UserIDResult(userID);
        callback.success(result);
    }

    public void connectedPeers(@NonNull Callback callback) {
        List<UUID> peers = bridgefy.connectedPeers();

        ConnectedPeersResult result = new ConnectedPeersResult(peers);
        callback.success(result);
    }

    /**
     * Secure Connection
     */

    public void establishSecureConnection(@NonNull EstablishSecureConnectionOptions options, @NonNull Callback callback) {
        @Nullable
        UUID userID = options.getUserID();

        if (userID == null) {
            callback.error(new Exception(MISSING_USER_IDENTIFIER));
            return;
        }

        bridgefy.establishSecureConnection(userID);

        callback.success();
    }

    public void fingerprint(@NonNull FingerprintOptions options, @NonNull Callback callback) {
        @Nullable
        UUID userID = options.getUserID();

        if (userID == null) {
            callback.error(new Exception(MISSING_USER_IDENTIFIER));
            return;
        }

        var fingerprint = bridgefy.fingerprint(userID);

        FingerprintResult result = new FingerprintResult(fingerprint);
        callback.success(result);
    }

    public void isFingerprintValid(@NonNull IsFingerprintValidOptions options, @NonNull Callback callback) {
        @Nullable
        UUID userID = options.getUserID();

        if (userID == null) {
            callback.error(new Exception(MISSING_USER_IDENTIFIER));
            return;
        }

        @Nullable
        byte[] fingerprint = options.getFingerprint();

        if (fingerprint == null) {
            callback.error(new Exception(MISSING_FINGERPRINT));
            return;
        }

        var isValid = bridgefy.isFingerprintValid(userID, fingerprint);

        IsFingerprintValidResult result = new IsFingerprintValidResult(isValid);
        callback.success(result);
    }

    /**
     * Payload
     */

    public void send(@NonNull SendOptions options, @NonNull Callback callback) {
        @Nullable
        byte[] data = options.getData();

        if (data == null) {
            callback.error(new Exception(MISSING_PAYLOAD));
            return;
        }

        @Nullable
        var transmissionMode = options.getTransmissionMode();

        if (transmissionMode == null) {
            callback.error(new Exception(MISSING_TRANSMISSION_MODE));
            return;
        }

        UUID messageID = bridgefy.send(data, transmissionMode);

        SendResult result = new SendResult(messageID);
        callback.success(result);
    }
}

File: ios/Sources/BridgefyPlugin/Bridgefy.swift
import Foundation
import Capacitor
import CoreBluetooth

@objc public class Bridgefy: NSObject {
    private let plugin: BridgefyPlugin
    private let config: BridgefyConfig?

    private let bridgefy: BridgefyController

    init(plugin: BridgefyPlugin, config: BridgefyConfig?) {
        self.plugin = plugin
        self.config = config

        self.bridgefy = BridgefyController(plugin)

        super.init()
    }

    deinit {
        stop {_ in}
    }

    /**
     * Initialize
     */

    @objc public func initialize(_ options: InitializeOptions, completion: @escaping (Error?) -> Void) {
        guard let apiKey = options.getAPIKey() ?? config?.getAPIKey() else {
            completion(CustomError.apiKeyMissing)
            return
        }

        let verboseLogging = options.getVerboseLogging() ?? config?.getVerboseLogging()

        do {
            try bridgefy.initialize(apiKey, verboseLogging)

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func isInitialized(completion: @escaping (Result?, Error?) -> Void) {
        let isInitialized = bridgefy.isInitialized()

        let result = IsInitializedResult(isInitialized)

        completion(result, nil)
    }

    @objc public func start(_ options: StartOptions, completion: @escaping (Error?) -> Void) {
        let userID = options.getUserID()

        let propagationProfile = options.getPropagationProfile()

        do {
            try bridgefy.start(userID, propagationProfile)

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func isStarted(completion: @escaping (Result?, Error?) -> Void) {
        let isStarted = bridgefy.isStarted()

        let result = IsStartedResult(isStarted)

        completion(result, nil)
    }

    @objc public func stop(completion: @escaping (Error?) -> Void) {
        do {
            try bridgefy.stop()

            completion(nil)
        } catch {
            completion(error)
        }
    }

    /**
     * License
     */

    @objc public func licenseExpirationDate(completion: @escaping (Result?, Error?) -> Void) {
        do {
            let licenseExpirationDate = try bridgefy.licenseExpirationDate()

            let result = LicenseExpirationDateResult(licenseExpirationDate)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    /**
     * Session
     */

    @objc public func destroySession(completion: @escaping (Error?) -> Void) {
        do {
            try bridgefy.destroySession()

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func currentUserID(completion: @escaping (Result?, Error?) -> Void) {
        do {
            let currentUserID = try bridgefy.currentUserID()

            let result = CurrentUserIDResult(currentUserID)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func connectedPeers(completion: @escaping (Result?, Error?) -> Void) {
        do {
            let peers = try bridgefy.connectedPeers()

            let result = ConnectedPeersResult(peers)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    /**
     * Secure Connection
     */

    @objc public func establishSecureConnection(_ options: EstablishSecureConnectionOptions, completion: @escaping (Error?) -> Void) {
        guard let userID = options.getUserID() else {
            completion(CustomError.userIDMissing)
            return
        }

        do {
            try bridgefy.establishSecureConnection(userID)

            completion(nil)
        } catch {
            completion(error)
        }
    }

    @objc public func fingerprint(_ options: FingerprintOptions, completion: @escaping (Result?, Error?) -> Void) {
        guard let userID = options.getUserID() else {
            completion(nil, CustomError.userIDMissing)
            return
        }

        do {
            let fingerprint = try bridgefy.fingerprint(userID)

            guard let fingerprint = fingerprint else {
                completion(nil, CustomError.fingerprintMissing)
                return
            }

            let result = FingerprintResult(fingerprint.scannable)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    @objc public func isFingerprintValid(_ options: IsFingerprintValidOptions, completion: @escaping (Result?, Error?) -> Void) {
        guard let userID = options.getUserID() else {
            completion(nil, CustomError.userIDMissing)
            return
        }

        guard let fingerprint = options.getFingerprint() else {
            completion(nil, CustomError.fingerprintMissing)
            return
        }

        do {
            let isValid = try bridgefy.isFingerprintValid(userID, fingerprint)

            let result = IsFingerprintValidResult(isValid)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    /**
     * Payload
     */

    @objc public func send(_ options: SendOptions, completion: @escaping (Result?, Error?) -> Void) {
        guard let data = options.getData() else {
            completion(nil, CustomError.payloadMissing)
            return
        }

        guard let transmissionMode = options.getTransmissionMode() else {
            completion(nil, CustomError.transmissionModeMissing)
            return
        }

        do {
            let messageID = try bridgefy.send(data, transmissionMode)

            let result = SendResult(messageID)

            completion(result, nil)
        } catch {
            completion(nil, error)
        }
    }

    /**
     * Permissions
     */

    @objc public func checkPermissions(completion: @escaping (Result?, Error?) -> Void) {
        let bluetoothState = switch CBCentralManager().state {
        case .unknown:
            "denied"
        case .resetting:
            "prompt"
        case .unsupported:
            "denied"
        case .unauthorized:
            "denied"
        case .poweredOff:
            "prompt"
        case .poweredOn:
            "granted"
        }
        let locationState = "granted"

        let result = PermissionsResult(bluetooth: bluetoothState, location: locationState)

        completion(result, nil)
    }

    @objc public func requestPermissions(_ options: RequestPermissionsOptions, completion: @escaping (Error?) -> Void) {
        guard let permissions = options.getPermissions() else {
            return
        }

        for permission in permissions {
            switch permission {
            case "bluetooth":
                guard let settingsUrl = URL(string: UIApplication.openSettingsURLString) else {
                    completion(CustomError.openSettingsError)
                    return
                }

                DispatchQueue.main.async {
                    if UIApplication.shared.canOpenURL(settingsUrl) {
                        UIApplication.shared.open(settingsUrl, completionHandler: { (success) in
                            if !success {
                                completion(CustomError.openSettingsError)
                                return
                            }
                        })
                    } else {
                        completion(CustomError.openSettingsError)
                        return
                    }
                }
            case "location":
                break
            default:
                break
            }
        }

        completion(nil)
    }

}
